python-package/xgboost/core.py

```diff
@@ -82,13 +82,11 @@ def from_pystr_to_cstr(data: Union[str, List[str]]) -> Union[bytes, ctypes.Array
     if isinstance(data, str):
         return bytes(data, "utf-8")
     if isinstance(data, list):
-        pointers: ctypes.Array[ctypes.c_char_p] = (ctypes.c_char_p * len(data))()
-        data_as_bytes = [bytes(d, "utf-8") for d in data]
-        pointers[:] = data_as_bytes  # type: ignore
+        data_as_bytes: List[bytes] = [bytes(d, "utf-8") for d in data]
+        pointers: ctypes.Array[ctypes.c_char_p] = (ctypes.c_char_p * len(data_as_bytes))(*data_as_bytes)
         return pointers
     raise TypeError()
 
-
```

Comment:
This is out of my wheelhouse somewhat, but to me it looks like we're initialising an empty ctypes.Array (`pointers`), then adding `data_as_bytes` to it which was creating a mismatch between the List[bytes] and ctypes.Array[ctypes.c_char_p]. I've just added the `data_as_bytes` into `pointers` at initialisation, does this make sense?

Reply:
To be honest, I don't quite understand it either. Need to look it up. ;-)